tech-specs/src/explore/design/dkg.md

```diff
@@ -1 +1,154 @@
+  broadcast their stake
+- `PartitionShares`: Once the block containing all of the `AnnounceStake`
+  messages are finalized, the set of participants is fixed and the
+  partition of the weight shares is added to the block.
+- `Deal`: This message contains the PVSS transcript from a validator
+  acting as a dealer
+- `Aggregate`: Combine the PVSS instances into a single one, thereby
+  producing the final key.
+
+In Ferveo, the DKG protocol is modelled as a state machine. This allows a
+validator to initialize an instance and drive it forward by reacting to
+each of the above message in the appropriate way. All necessary state
+is persisted by this state machine. We describe the  appropriate action for
+each of the above transactions.
+
+ - `StartDKG`: Initialize a fresh instace of the DKG protocol from Ferveo.
+   Then send out an `AnnounceStake` transaction.
+ - `AnnounceStake`: These simply need to be verified. The current block
+   proposer adds a `PartitionShares` message to the block after the block starting
+   the DKG instance. It is based off all the `AnnounceStake` transactions
+   which will also be included in the next block (for verification).
+ - `PartitionShares`: Validators verify the partitions and weights and
+   should extract their weight shares and store them in the DKG state machine.
+   Then they should send out `Deal` transactions.
+ - `Deal`: These need to be verified. The current block proposer checks if 
+   enough PVSS instances have been reached to meet the security threshold.
+   If so, it adds a `Aggregate` transaction to the block. Otherwise, the
+   next block proposer will take on the job. 
+ - `Aggregate`: This simply needs to be verified and the resulting key added
+   to the DKG instance. 
+
+Most of the above transactions occur as a reaction to other messages
+and/or data in the proposed block header. The three exceptions are the 
+`StartDKG`, `PartitionShares`, and `Aggregate` transactions. These must be 
+made by the current block proposer if their current DKG state machine is in the 
+appropriate state and/or the appropriate conditions are met. Thus block 
+proposers must check the state machines of their active DKG instances.
+
+## ABCI++
+
+The creation of the public key and the shared secrets requires ABCI++ as we
+need control over block preparation as well extra data from the Vote Extension
+phase. 
+
+We list the phase of ABCI++ in which each of the above actions takes place.
+ - `StartDKG`: This is added to the block by the current proposer as part of 
+    the Prepare Proposal phase.
+ - `AnnounceStake`: These are sent out by validators during the Vote
+    Extension phase.
+ - `PartitionShares`: This also happens in the Vote Extension phase and the
+    result will be included in the next block's header.
+ - `Deal`: Validators should verify the weight shares included in the block 
+    header. If validation fails, the DKG instance fails.  Otherwise, `Deal` 
```

Comment:
Since partition shares should happen implicitly, I think we don't have to worry about its failure